File: crates/batchalign-transform/src/tokenizer_realign.rs
//! Tokenizer realignment: merge Stanza tokens back to original CHAT words.
//!
//! When Stanza's neural tokenizer runs, it may re-split compound words
//! (e.g. "ice-cream" → `["ice", "-", "cream"]`). This module merges those
//! spurious splits back, preserving the 1-to-1 mapping between CHAT words
//! and Stanza tokens.
//!
//! # Architecture
//!
//! This module is the shared implementation used by both:
//! - The PyO3 bridge (`batchalign-core`) — called from `align_tokens` pyfunction
//! - The standalone Rust server (`batchalign-server`) — called directly
//!
//! The PyO3 crate provides only a thin wrapper that converts `Vec<PatchedToken>`
//! to Python objects (`str` or `(str, bool)` tuples).
//!
//! # MWT Hint Convention
//!
//! Stanza's `tokenize_postprocessor` uses a tuple convention:
//!
//! - `(text, True)`  — MWT: let the MWT processor expand (e.g. "don't" → do + n't)
//! - `(text, False)` — NOT an MWT: suppress expansion (e.g. merged "ice-cream")
//! - plain string    — let Stanza's model decide
//!
//! [`PatchedToken`] encodes this convention at the Rust↔Python boundary.
//! The [`Hint`](PatchedToken::Hint) variant is emitted only when the
//! character-DP merges multiple Stanza tokens into one and the merged text
//! looks like an English contraction ([`is_contraction`]). All other cases
//! produce [`Plain`](PatchedToken::Plain).
//!
//! # Provenance
//!
//! The character-position mapping algorithm replaces batchalign2's DP-based
//! `tokenizer_processor` (ud.py:610-700). Per-language MWT override rules
//! (BA2 `ud.py:662-695`) were audited and retired on 2026-04-21 after paired
//! probes showed every rule was dormant, redundant, or harmful; see
//! `book/src/reference/languages/{french,italian,portuguese,dutch}.md` for
//! the audit records. The character-DP alone satisfies the morphotag 1-to-1
//! invariant for the five previously-patched languages.
//!
//! See: `book/src/reference/morphotag-migration-audit.md` Section 6.

/// Token produced by [`align_tokens`] at the Rust↔Python boundary.
///
/// Encodes Stanza's tokenize-postprocessor MWT hint convention:
/// - `Plain(text)` — no hint, let Stanza's model decide
/// - `Hint(text, true)` — force MWT expansion
/// - `Hint(text, false)` — suppress MWT expansion
#[derive(Debug, Clone, PartialEq)]
pub enum PatchedToken {
    /// Plain string — no MWT hint.
    Plain(String),
    /// MWT hint tuple: `(text, should_expand)`.
    Hint(String, bool),
}

impl PatchedToken {
    /// Extract the text content regardless of variant.
    pub fn text(&self) -> &str {
        match self {
            PatchedToken::Plain(s) | PatchedToken::Hint(s, _) => s,
        }
    }
}

// ─── Contraction detection ──────────────────────────────────────────────────

/// Whether a merged token should be flagged as an English MWT contraction.
///
/// Replicates batchalign2's `ud.py` lines 680-685:
///   - Token contains `'`
///   - Language is English (`alpha2 == "en"`)
///   - The prefix before the first `'` is NOT `"o"` (excludes o'clock, o'er)
pub fn is_contraction(text: &str, alpha2: &str) -> bool {
    if !text.contains('\'') {
        return false;
    }
    if alpha2 != "en" {
        return false;
    }
    // Exclude o'clock, o'er, etc. — prefix before first apostrophe is "o"
    if let Some(prefix) = text.split('\'').next()
        && prefix.trim().to_lowercase() == "o"
    {
        return false;
    }
    true
}

// ─── Core alignment algorithm ───────────────────────────────────────────────

/// Align Stanza tokenizer output back to original CHAT words.
///
/// Uses character-position mapping: the concatenated characters of Stanza tokens
/// must equal the concatenated characters of original words (no reordering).
/// If they don't match, the function returns the Stanza tokens unchanged as
/// `PatchedToken::Plain` values.
///
/// After merging, language-specific MWT patches are applied (French, Italian,
/// Portuguese, Dutch).
///
/// # Arguments
///
/// - `original_words` — CHAT words (may contain shortening parens which are stripped)
/// - `stanza_tokens` — tokens from Stanza's neural tokenizer
/// - `alpha2` — ISO-639-1 language code (e.g. `"en"`, `"fr"`)
pub fn align_tokens(
    original_words: &[String],
    stanza_tokens: &[String],
    alpha2: &str,
) -> Vec<PatchedToken> {
    if stanza_tokens.is_empty() || original_words.is_empty() {
        return stanza_tokens
            .iter()
            .map(|t| PatchedToken::Plain(t.clone()))
            .collect();
    }

    // Clean original words (strip CHAT shortening parens)
    let cleaned: Vec<String> = original_words
        .iter()
        .map(|w| w.replace(['(', ')'], ""))
        .collect();

    // Build per-character maps: which word / which token does each char belong to?
    let mut ref_chars: Vec<usize> = Vec::new();
    for (word_idx, word) in cleaned.iter().enumerate() {
        for _ in word.chars() {
            ref_chars.push(word_idx);
        }
    }

    let mut tok_chars: Vec<usize> = Vec::new();
    for (tok_idx, tok) in stanza_tokens.iter().enumerate() {
        for _ in tok.chars() {
            tok_chars.push(tok_idx);
        }
    }

    // Character content must match — bail out if they don't
    let ref_str: String = cleaned.iter().flat_map(|w| w.chars()).collect();
    let tok_str: String = stanza_tokens.iter().flat_map(|t| t.chars()).collect();
    if ref_str != tok_str {
        return stanza_tokens
            .iter()
            .map(|t| PatchedToken::Plain(t.clone()))
            .collect();
    }

    // Build word->token mapping: for each original word, which token indices span it?
    let num_words = cleaned.len();
    let mut word_to_tokens: Vec<Vec<usize>> = vec![vec![]; num_words];
    for char_idx in 0..ref_chars.len() {
        let w = ref_chars[char_idx];
        let t = tok_chars[char_idx];
        if word_to_tokens[w].last() != Some(&t) {
            word_to_tokens[w].push(t);
        }
    }

    // 1-to-1 case: one Stanza token per CHAT word, no merge needed.
    if word_to_tokens.iter().all(|toks| toks.len() == 1) {
        return stanza_tokens
            .iter()
            .map(|t| PatchedToken::Plain(t.clone()))
            .collect();
    }

    // Merge case: at least one CHAT word spans multiple Stanza tokens.
    // Emit PatchedToken::Hint only when the merged text looks like an
    // English contraction (so Stanza's MWT processor expands it back).
    let num_tokens = stanza_tokens.len();
    let mut seen = vec![false; num_tokens];
    let mut tokens: Vec<PatchedToken> = Vec::with_capacity(num_words);

    for toks in &word_to_tokens {
        let unseen: Vec<usize> = toks.iter().copied().filter(|&i| !seen[i]).collect();
        if unseen.is_empty() {
            continue;
        }
        if unseen.len() == 1 {
            tokens.push(PatchedToken::Plain(stanza_tokens[unseen[0]].clone()));
        } else {
            let merged: String = unseen.iter().map(|&i| stanza_tokens[i].as_str()).collect();
            let is_contr = is_contraction(&merged, alpha2);
            tokens.push(PatchedToken::Hint(merged, is_contr));
        }
        for &i in &unseen {
            seen[i] = true;
        }
    }

    // Append any unmapped tokens (shouldn't happen with correct alignment).
    for (tok_idx, tok) in stanza_tokens.iter().enumerate() {
        if !seen[tok_idx] {
            tokens.push(PatchedToken::Plain(tok.clone()));
        }
    }

    tokens
}

// ─── Tests ──────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    // ── is_contraction ────────────────────────────────────────────────

    #[test]
    fn test_english_contraction_detected() {
        assert!(is_contraction("don't", "en"));
        assert!(is_contraction("I'm", "en"));
        assert!(is_contraction("Claus'", "en"));
    }

    #[test]
    fn test_english_oclock_not_contraction() {
        assert!(!is_contraction("o'clock", "en"));
        assert!(!is_contraction("O'er", "en"));
    }

    #[test]
    fn test_non_english_not_contraction() {
        assert!(!is_contraction("l'homme", "fr"));
        assert!(!is_contraction("d'água", "pt"));
    }

    #[test]
    fn test_no_apostrophe_not_contraction() {
        assert!(!is_contraction("hello", "en"));
    }

    // ── align_tokens ─────────────────────────────────────────────────

    #[test]
    fn test_empty_inputs() {
        let result = align_tokens(&[], &[], "en");
        assert!(result.is_empty());
    }

    #[test]
    fn test_one_to_one_mapping() {
        let words = vec!["hello".into(), "world".into()];
        let tokens = vec!["hello".into(), "world".into()];
        let result = align_tokens(&words, &tokens, "en");
        assert_eq!(
            result,
            vec![
                PatchedToken::Plain("hello".into()),
                PatchedToken::Plain("world".into()),
            ]
        );
    }

    #[test]
    fn test_merge_compound() {
        // Stanza splits "ice-cream" into ["ice", "-", "cream"]
        let words = vec!["ice-cream".into()];
        let tokens = vec!["ice".into(), "-".into(), "cream".into()];
        let result = align_tokens(&words, &tokens, "en");
        assert_eq!(result, vec![PatchedToken::Hint("ice-cream".into(), false)]);
    }

    #[test]
    fn test_english_contraction_merge() {
        // Stanza splits "don't" into ["do", "n't"]
        let words = vec!["don't".into()];
        let tokens = vec!["do".into(), "n't".into()];
        let result = align_tokens(&words, &tokens, "en");
        assert_eq!(result, vec![PatchedToken::Hint("don't".into(), true)]);
    }

    #[test]
    fn test_character_mismatch_passthrough() {
        let words = vec!["hello".into()];
        let tokens = vec!["goodbye".into()];
        let result = align_tokens(&words, &tokens, "en");
        assert_eq!(result, vec![PatchedToken::Plain("goodbye".into())]);
    }

    #[test]
    fn test_shortening_parens_stripped() {
        // CHAT shortening: "(be)cause" → characters "because"
        let words = vec!["(be)cause".into()];
        let tokens = vec!["because".into()];
        let result = align_tokens(&words, &tokens, "en");
        assert_eq!(result, vec![PatchedToken::Plain("because".into())]);
    }

    // ── align_tokens: cross-language passthrough (no per-language rules) ──

    #[test]
    fn test_english_passthrough_no_patches() {
        let words = vec!["the".into(), "dog".into()];
        let tokens = vec!["the".into(), "dog".into()];
        let result = align_tokens(&words, &tokens, "en");
        assert_eq!(
            result,
            vec![
                PatchedToken::Plain("the".into()),
                PatchedToken::Plain("dog".into()),
            ]
        );
    }
}

File: batchalign/inference/_tokenizer_realign.py
"""Tokenizer realignment for tokenize_no_ssplit mode.

When Stanza's neural tokenizer runs (``tokenize_no_ssplit=True``), it may
split compound words like "ice-cream" into multiple tokens ("ice", "-",
"cream").  This module provides a ``tokenize_postprocessor`` callback that
merges such spurious splits back, preserving the 1:1 mapping between CHAT
words and Stanza tokens.

MWT hints (True/False tuples)
------------------------------
Stanza's ``tokenize_postprocessor`` uses a tuple convention::

    (text, True): MWT: let the MWT processor expand (e.g. "don't" → do + n't)
    (text, False), NOT an MWT: suppress expansion (e.g. merged "ice-cream")
    plain string: let Stanza's model decide (equivalent to model's own choice)

This module replicates Python master's ``tokenizer_processor`` logic:

* **Default (all languages)**: merged spurious splits → ``(text, False)``
  Prevents a merge like "ice-cream" from being expanded again by the MWT model.
* **English contractions**: merged text that contains ``'``, *unless* the
  prefix before the first ``'`` is ``"o"`` (e.g. o'clock, o'er) → ``(text, True)``
  Allows "don't", "Claus'" etc. to be handled by Stanza's MWT model.

This matches the Python master rules (``ud.py`` lines 680-685) exactly.

Thread safety: :class:`TokenizerContext` uses ``threading.local()`` to store
``original_words`` per-thread.  On free-threaded Python (3.14t+), multiple
threads can call ``nlp()`` concurrently without racing on the context.  On
regular Python, the ``nlp_lock`` in the batch callback still serializes
access, so the thread-local is effectively a single-thread property.
"""

from __future__ import annotations

import logging
import threading
from collections.abc import Callable
from typing import TypeAlias

from batchalign.inference._italian_mwt import (
    ItalianMwtPolicy,
    MwtExpansion,
    could_be_enclisis,
    decide_expansion,
)

L = logging.getLogger("batchalign")


TokenizerToken: TypeAlias = str | tuple[str, bool]


class TokenizerContext:
    """Thread-safe context shared between the batch callback and the postprocessor.

    Uses ``threading.local()`` so each thread's ``original_words`` is
    independent: required for free-threaded Python where multiple threads
    call ``nlp()`` concurrently on the same Pipeline.
    """

    def __init__(self) -> None:
        self._local = threading.local()

    @property
    def original_words(self) -> list[list[str]]:
        return getattr(self._local, "original_words", [])

    @original_words.setter
    def original_words(self, value: list[list[str]]) -> None:
        self._local.original_words = value


def make_tokenizer_postprocessor(
    ctx: TokenizerContext,
    alpha2: str = "",
    italian_policy: ItalianMwtPolicy | None = None,
) -> Callable[[list[list[TokenizerToken]]], list[list[TokenizerToken]]]:
    """Create a ``tokenize_postprocessor`` callback for ``stanza.Pipeline``.

    The returned callable has the signature Stanza expects::

        postprocessor(tokenized_batch: list[list]) -> list[list]

    where each inner list is the tokens for one sentence (paragraph).

    Parameters
    ----------
    ctx:
        Mutable context updated before each ``nlp()`` call with the original
        CHAT words for the current batch.
    alpha2:
        ISO-639-1 language code (e.g. ``"en"``, ``"fr"``).  Used to decide
        whether merged tokens should be flagged as MWT contractions.
    italian_policy:
        Inputs for the Italian single-word multi-word-token decision. Applied
        only when ``alpha2 == "it"``; ``None`` leaves Stanza's own behavior
        untouched.
    """

    def postprocessor(
        tokenized_batch: list[list[TokenizerToken]],
    ) -> list[list[TokenizerToken]]:
        if not ctx.original_words:
            return tokenized_batch

        result: list[list[TokenizerToken]] = []
        for sent_idx, sent_tokens in enumerate(tokenized_batch):
            if sent_idx < len(ctx.original_words):
                original = ctx.original_words[sent_idx]
                result.append(_realign_sentence(sent_tokens, original, alpha2))
            else:
                result.append(sent_tokens)

        # Applied to the WHOLE batch, not per sentence: the decision needs a
        # second Stanza pass over the candidate words, and doing that once per
        # batch instead of once per utterance is the difference between one
        # extra model call and several hundred.
        if alpha2 == "it" and italian_policy is not None:
            result = _apply_italian_mwt_policy(result, italian_policy)
        return result

    return postprocessor


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _conform(token: TokenizerToken) -> str:
    """Extract text from a Stanza token (string or tuple)."""
    if isinstance(token, tuple):
        return str(token[0])
    return str(token)


def _is_contraction(text: str, alpha2: str) -> bool:
    """Return True if *text* should be flagged as an MWT contraction.

    Replicates Python master's ``tokenizer_processor`` English contraction rule
    (``ud.py`` lines 680-685)::

        (("en" in lang) and matches_in(i, "'") and
         not (len(conform(i).split("'")) > 1 and
              conform(i).split("'")[0].strip() == "o"))

    Returns ``True`` only for English tokens that contain an apostrophe and
    whose prefix before the first ``'`` is not ``"o"`` (which would be forms
    like o'clock, o'er, o'er the top, etc.).

    All other tokens return ``False``, meaning the MWT model will NOT try to
    expand them (suppresses spurious re-expansion of merged words).
    """
    if "'" not in text:
        return False
    if alpha2 != "en":
        return False
    # Exclude o'clock, o'er, o'er, etc. (prefix before first apostrophe is "o")
    parts = text.split("'")
    if len(parts) >= 2 and parts[0].strip().lower() == "o":
        return False
    return True


def _realign_sentence(
    stanza_tokens: list[TokenizerToken],
    original_words: list[str],
    alpha2: str = "",
) -> list[TokenizerToken]:
    """Merge Stanza tokens that map to the same original CHAT word.

    Delegates to ``batchalign_core.align_tokens()`` (Rust) for the
    character-position mapping algorithm. Per-language MWT override rules
    were retired on 2026-04-21 after a paired empirical audit, the
    character-DP alone satisfies the morphotag 1-to-1 invariant for all
    previously-patched languages.

    Stanza may return tokens with embedded spaces (rare edge case). These
    are flattened before passing to Rust so the character sequences match.

    Returned items may be plain strings or ``(text, bool)`` tuples, matching
    Stanza's postprocessor contract for MWT expansion hints.

    MWT hint preservation
    ---------------------
    Stanza's tokenizer natively emits ``(text, True)`` tuples for tokens
    it wants its MWT processor to expand (English contractions, French
    elisions, etc.). The Rust ``align_tokens`` function operates on raw
    strings for its char-DP algorithm, so tuple hints are temporarily
    erased when we flatten. When no merging actually happened (1:1
    mapping between Stanza tokens and the aligner's output), we overlay
    Stanza's original tuples back onto the aligned output so MWT
    expansion still fires downstream. Without this overlay, Stanza's MWT
    processor sees only plain strings and silently skips expansion, the
    direct cause of the 2026-04-13 Preserve-mode regression.
    """
    if not stanza_tokens or not original_words:
        return stanza_tokens

    # Flatten tokens that Stanza may have returned with embedded spaces
    flat_tokens: list[str] = []
    for tok in stanza_tokens:
        text = _conform(tok)
        parts = text.split(" ")
        flat_tokens.extend(parts if len(parts) > 1 else [text])

    from batchalign_core import align_tokens
    merged = align_tokens(original_words, flat_tokens, alpha2)

    # Restore Stanza's own MWT hint tuples wherever a merged/aligned token
    # still corresponds 1:1 to one original Stanza token. This matters even
    # when another token elsewhere in the sentence was merged (e.g.
    # "mm" + "-" + "hmm" -> "mm-hmm"): the unaffected English contraction
    # later in the sentence still needs its native (text, True) hint so the
    # MWT processor expands it.
    restored: list[TokenizerToken] = []
    orig_idx = 0
    for new in merged:
        target = _conform(new)
        start = orig_idx
        buf = ""
        while orig_idx < len(stanza_tokens) and len(buf) < len(target):
            buf += _conform(stanza_tokens[orig_idx])
            orig_idx += 1
            if buf == target:
                break

        # If we cannot reconcile the aligned token sequence back onto the
        # original Stanza sequence, return the aligner output unchanged.
        if buf != target:
            return merged

        if isinstance(new, tuple):
            restored.append(new)
            continue

        if (
            orig_idx - start == 1
            and isinstance(stanza_tokens[start], tuple)
            and _conform(stanza_tokens[start]) == target
        ):
            restored.append(stanza_tokens[start])
        else:
            restored.append(new)

    if orig_idx != len(stanza_tokens):
        return merged

    return restored


def _apply_italian_mwt_policy(
    batch: list[list[TokenizerToken]],
    policy: ItalianMwtPolicy,
) -> list[list[TokenizerToken]]:
    """Suppress Italian MWT expansion where Stanza's proposed split is not real.

    Stanza's Italian MWT over-splits, inventing verbs. Context-free it turns
    ``cavallo`` into *cava* + *lo* and ``attenzione`` into *attenzi* + *ne*; in
    full sentences it turns the article ``la`` into *il* + *i* and ``mozzarella``
    into *mozzar* + *la*. Both are the same defect, so both are judged here by
    the same rule. What counts as a genuine Italian multi-word token, and the
    measurements behind it, live in ``batchalign/inference/_italian_mwt.py``.

    Candidates are exactly the tokens Stanza itself marked ``(text, True)``,
    which is its documented way of saying "the MWT processor should expand
    this". Reading that marker rather than guessing which words might be
    multi-word tokens is what makes the pass both correct in every context and
    cheap: a typical sentence marks two tokens out of eleven, and the rest are
    never probed.

    Emitting ``(text, False)`` is the same protocol in reverse: do not expand.
    """
    lexicon = policy.lexicon()
    if lexicon is None:
        # A check that did not run must not change the analysis.
        L.debug("Italian MWT policy inactive: lexicon unavailable")
        return batch

    # Two kinds of candidate, judged by the same rule:
    #  - tokens Stanza already marked `(text, True)`, which may be over-splits;
    #  - tokens it left alone that look lexically like enclisis, which may be
    #    the opposite defect, where it declines to split a real imperative and
    #    invents a verb for the whole surface (`aprilo` -> `verb|aprilare`).
    candidates: dict[tuple[int, int], str] = {}
    for sent_idx, tokens in enumerate(batch):
        for tok_idx, token in enumerate(tokens):
            text = _conform(token)
            already_marked = isinstance(token, tuple) and token[1]
            if already_marked or could_be_enclisis(text, lexicon):
                candidates[(sent_idx, tok_idx)] = text

    if not candidates:
        return batch

    # One probe call for the whole batch. It gets whole utterances, not the
    # candidate words alone, because Stanza's MWT is context-sensitive: `hai`
    # in isolation is left whole but becomes *ha* + *i* mid-sentence, so an
    # isolated probe would report nothing to judge and the over-split would
    # pass straight through.
    #
    # Every candidate is force-hinted. That costs nothing for the ones Stanza
    # was already going to expand (the hint is what it emitted itself), and it
    # is the only way to see the split it declines to make.
    utterances = [[_conform(token) for token in tokens] for tokens in batch]
    per_sentence = policy.propose_splits(utterances, frozenset(candidates.values()))
    if per_sentence is None:
        L.debug("Italian MWT policy inactive: split probe unavailable")
        return batch

    adjusted = [list(tokens) for tokens in batch]
    for (sent_idx, tok_idx), text in candidates.items():
        if sent_idx >= len(per_sentence):
            continue
        split = per_sentence[sent_idx].get(text)
        if split is None or len(split) < 2:
            # Nothing to judge: leave the token exactly as Stanza left it rather
            # than asserting an expansion that has no proposed shape.
            continue
        if decide_expansion(text, split, lexicon) is MwtExpansion.ALLOW:
            adjusted[sent_idx][tok_idx] = (text, True)
        else:
            adjusted[sent_idx][tok_idx] = (text, False)
    return adjusted
